botocore/handlers.py

```diff
@@ -282,9 +282,11 @@ def add_expect_header(model, params, **kwargs):
         return
     if 'body' in params:
         body = params['body']
-        if hasattr(body, 'read'):
+        size = getattr(body, "_size", None)
```

Comment:
This unfortunately won't work because it relies on a private variable that's not always present. This change is a fundamental rework of how we compute this which is breaking. Objects that currently implement `read` without `_size` will stop emitting a 100. This also introduces a forward compatibility risk, that if an object ever removes, renames, or changes data types of the private `_size` variable, this logic fails unexpectedly.

If we introduce a solution, we need it to rely solely on public interfaces. It also needs to operate without reading the body, because some inputs are not rewindable and the data will be lost. This is what introduces the complexity that have prevented previous fixes.

We agree with what the SDK _should_ do in this case, but given it's not a general HTTP client we've scoped it to accepted behavior within the AWS ecosystem. Otherwise, we wouldn't be able to support critical features in S3. Long term, we're looking at more flexible solutions for other platforms, but have nothing concrete at this time.